File: scripts/get_metrics.py
from datetime import datetime
from faker import Faker
import sys

# from plot_data import plot_data

# Check if a day is a saturday
def is_saturday(date_string):
    # Parse the date string into a datetime object
    date_object = datetime.strptime(date_string, "%Y-%m-%d")
    
    # Check if the weekday is Saturday (5 represents Saturday)
    return date_object.weekday() == 5

# sorts the scores acheieved by date
def process_scores(input_text, cutoff_date):
    lines = input_text.splitlines()
    scores_struct = []
    for line in lines:
        parts = line.split(",")
        name = parts[0].strip()  # Remove leading/trailing whitespaces
        date = parts[1].strip()  # Remove leading/trailing whitespaces
        time = int(parts[2].strip())  # Remove leading/trailing whitespaces and convert to integer
        if cutoff_date <= date: # filter by cutoff date input argument, 
            if time <= 1200: # remove times that are over 20 minutes
                scores_struct.append((name, date, time))

    sorted_list = sorted(scores_struct, key=lambda x: x[1])
    return sorted_list

def anonymize_names(sorted_list):
    Faker.seed(1)
    faker = Faker()
    dict_names = {name: faker.name() for name, _, _ in sorted_list}
    #NOTE: Could hardcode names here if we wanted to
    anonymized_sorted_list = [(dict_names[name], date, time) for name, date, time in sorted_list]
    return anonymized_sorted_list

# calculate each person's average non-saturday and saturday time
def get_average_time(scores, cutoff_date):
    # Open a new file for writing output
    with open("../text_files/averages.txt", "w") as output_file:
        dict = {}
        for score in scores:
            name, date, time = score
            
            if is_saturday(date):
                # third and forth list spots are for saturdays
                if name in dict:
                    original_tuple = dict[name]
                    dict[name] = (original_tuple[0], original_tuple[1], original_tuple[2] + time, original_tuple[3] + 1)
                else:
                    dict[name] = (0, 0, time, 1)
            else:
                # first and second list spots are for non-saturdays
                if name in dict:
                    original_tuple = dict[name]
                    dict[name] = (original_tuple[0] + time, original_tuple[1] + 1, original_tuple[2], original_tuple[3])
                else:
                    dict[name] = (time, 1, 0, 0)

        averages_tuples = []
        for name, tuple in dict.items():
            # get average times for saturdays and non-saturdays, ensuring we don't divide by zero
            averages_tuples.append((name, round(tuple[0] / tuple[1] if tuple[1] != 0 else -1, 2), tuple[1], 
                                    round(tuple[2] / tuple[3] if tuple[3] != 0 else -1, 2), tuple[3]))

        sorted_list = sorted(averages_tuples, key=lambda x: x[1])

        # headers
        if cutoff_date:
            output_file.write("Cutoff day: " + cutoff_date + "\n")
        output_file.write("Name, Average Time, Number of Times\n")

        for score in sorted_list:
            output_file.write(f"{score[0]}, {score[1]}, {score[2]}\n")

        output_file.write("\nSaturdays:\n")

        # Sort Saturdays
        sorted_list = sorted(averages_tuples, key=lambda x: x[3])

        for score in sorted_list:
            output_file.write(f"{score[0]}, {score[3]}, {score[4]}\n")

# calculate each person's average position compared to others' times
def find_average_place(individualsDict):
    with open("../text_files/average_placements.txt", "w") as output_file:
        output_file.write("Average placements per person:")
        personAveragesList = []

        # get each person's average
        for person, placesList in individualsDict.items():
            average = round(sum(placesList) / len(placesList), 2)
            personAveragesList.append((person, average))

        # sort the averages
        sorted_list = sorted(personAveragesList, key=lambda x: x[1])
        for person, average in sorted_list:
            output_file.write(f"\n{person}, {average}")

# find the number of times a person has come first for that day
def find_number_of_firsts(individualsDict):
    with open("../text_files/first_places.txt", "w") as output_file:
        output_file.write("Number of first place finishes per person:")
        firstPlacesList = []
        # count each person's number of times being first
        for person, placesList in individualsDict.items():
            firstPlaces = placesList.count(1)
            firstPlacesList.append((person, firstPlaces))
        
        # sort the number of times someone got first, most at the top
        sorted_list = reversed(sorted(firstPlacesList, key=lambda x: x[1]))
        for person, firstPlaces in sorted_list:
            output_file.write(f"\n{person}, {firstPlaces}")

# processes peoples' placement for each day to make other calculations easier
def analyze_placement_info(scores, cutoff_date):
    individualsDict = {} # key is name, value is list of placements
    dayScoresList = []
    currentDate = cutoff_date

    for idx, score in enumerate(scores):
        name, date, time = score

        # add to current day which is being analyzed
        if date == currentDate and idx != len(scores)-1:
            dayScoresList.append((name, time))
        # we have gone through all the stats for today
        else:
            sorted_list = sorted(dayScoresList, key=lambda x: x[1])
            for place, tuple in enumerate(sorted_list):
                nameInTuple, score = tuple
                
                # if was a twin
                while place > 0 and score == sorted_list[place - 1][1]: # loop cause more than two people could've tied
                    # then give them the same placement as their twin, that just happened to be sorted first
                    place -= 1
                
                if nameInTuple in individualsDict:
                    individualsDict[nameInTuple].append(place+1)
                else:
                    individualsDict[nameInTuple] = [place+1]

            currentDate = date
            dayScoresList = [(name, time)]

    find_average_place(individualsDict)
    find_number_of_firsts(individualsDict)

def count_times_occurences(sorted_scores, cutoff_date):
    time_occurences_dict = {}
    for score in sorted_scores:
        time = score[2]
        if time in time_occurences_dict:
            time_occurences_dict[time] += 1
        else:
            time_occurences_dict[time] = 1
    with open("../text_files/time_occurences.txt", "w") as output_file:
        output_file.write("Number of occurences for each time (cutoff date is "+ cutoff_date + "):\n")
        for time in dict(sorted(time_occurences_dict.items())):
            output_file.write("Time: " + str(time) + "   Occurence: " + str(time_occurences_dict[time]) + "\n")
    with open("../text_files/most_common_times.txt", "w") as output_file:
        output_file.write("Times sorted by which is most common (cutoff date is "+ cutoff_date + "):\n")
        for time in dict(sorted(time_occurences_dict.items(), key=lambda item: (item[1], item[0]), reverse=True)):
            output_file.write("Occurence: " + str(time_occurences_dict[time]) + "    Time: " + str(time) +  "\n")

def main(cutoff_date = "2023-02-21"):
    # Read the input file
    with open("../text_files/cleaned_data.txt", "r") as file:
        text = file.read()

    sorted_scores = process_scores(text, cutoff_date) # format is [[name, date, time], ...]
    anon_sorted_scores = anonymize_names(sorted_scores)
    get_average_time(anon_sorted_scores, cutoff_date)
    analyze_placement_info(anon_sorted_scores, cutoff_date)
    count_times_occurences(anon_sorted_scores, cutoff_date)
    # plot_data(anon_sorted_scores)


if __name__ == "__main__":
    # Check if command-line arguments were provided
    if len(sys.argv) >= 2:
        arg1 = sys.argv[1]
        main(arg1)
    else:
        # Call main with default values
        main()

File: my-app/src/scripts/get_metrics.js
const fs = require('fs').promises;
const path = require('path');
const { faker } = require('@faker-js/faker');

// Utility functions
const isSaturday = (dateString) => new Date(dateString).getDay() === 6;

const roundToTwoDecimals = (num) => Math.round(num * 100) / 100;

const writeToFile = async (filename, content) => {
    const filePath = path.join(__dirname, '..', 'text_files', filename);
    await fs.writeFile(filePath, content);
};

// Core functions
const processScores = (inputText, cutoffDate) => {
    return inputText.split('\n')
        .map(line => line.split(',').map(item => item.trim()))
        .filter(([, date, time]) => date >= cutoffDate && parseInt(time) <= 1200)
        .map(([name, date, time]) => [name, date, parseInt(time)])
        .sort((a, b) => a[1].localeCompare(b[1]));
};

const anonymizeNames = (sortedList) => {
    faker.seed(1);
    const nameMap = new Map();
    return sortedList.map(([name, date, time]) => [
        nameMap.get(name) || nameMap.set(name, faker.name.fullName()).get(name),
        date,
        time
    ]);
};

const calculateAverageTimes = (scores) => {
    const timeData = scores.reduce((acc, [name, date, time]) => {
        if (!acc[name]) acc[name] = { regular: [], saturday: [] };
        (isSaturday(date) ? acc[name].saturday : acc[name].regular).push(time);
        return acc;
    }, {});

    return Object.entries(timeData).map(([name, { regular, saturday }]) => ({
        name,
        regularAvg: regular.length ? roundToTwoDecimals(regular.reduce((a, b) => a + b) / regular.length) : -1,
        regularCount: regular.length,
        saturdayAvg: saturday.length ? roundToTwoDecimals(saturday.reduce((a, b) => a + b) / saturday.length) : -1,
        saturdayCount: saturday.length
    }));
};

const analyzeAverageTimes = async (averageTimes, cutoffDate) => {
    const regularSorted = averageTimes.sort((a, b) => a.regularAvg - b.regularAvg);
    const saturdaySorted = averageTimes.sort((a, b) => a.saturdayAvg - b.saturdayAvg);

    const output = `Cutoff day: ${cutoffDate}
Name, Average Time, Number of Times
${regularSorted.map(({ name, regularAvg, regularCount }) => `${name}, ${regularAvg}, ${regularCount}`).join('\n')}

Saturdays:
${saturdaySorted.map(({ name, saturdayAvg, saturdayCount }) => `${name}, ${saturdayAvg}, ${saturdayCount}`).join('\n')}`;

    await writeToFile('averages.txt', output);
};

const analyzePlacementInfo = (scores) => {
    const placements = scores.reduce((acc, [name, date, time], _, arr) => {
        if (!acc[date]) acc[date] = arr.filter(score => score[1] === date).sort((a, b) => a[2] - b[2]);
        const place = acc[date].findIndex(score => score[2] === time) + 1;
        if (!acc.individuals[name]) acc.individuals[name] = [];
        acc.individuals[name].push(place);
        return acc;
    }, { individuals: {} });

    return placements.individuals;
};

const findAveragePlace = async (individualsDict) => {
    const averagePlacements = Object.entries(individualsDict).map(([person, places]) => ({
        person,
        average: roundToTwoDecimals(places.reduce((a, b) => a + b) / places.length)
    })).sort((a, b) => a.average - b.average);

    const output = `Average placements per person:
${averagePlacements.map(({ person, average }) => `${person}, ${average}`).join('\n')}`;

    await writeToFile('average_placements.txt', output);
};

const findNumberOfFirsts = async (individualsDict) => {
    const firstPlaces = Object.entries(individualsDict).map(([person, places]) => ({
        person,
        firsts: places.filter(place => place === 1).length
    })).sort((a, b) => b.firsts - a.firsts);

    const output = `Number of first place finishes per person:
${firstPlaces.map(({ person, firsts }) => `${person}, ${firsts}`).join('\n')}`;

    await writeToFile('first_places.txt', output);
};

const countTimesOccurrences = async (sortedScores, cutoffDate) => {
    const occurrences = sortedScores.reduce((acc, [, , time]) => {
        acc[time] = (acc[time] || 0) + 1;
        return acc;
    }, {});

    const byTime = Object.entries(occurrences)
        .sort(([a], [b]) => parseInt(a) - parseInt(b))
        .map(([time, count]) => `Time: ${time}   Occurrence: ${count}`)
        .join('\n');

    const byOccurrence = Object.entries(occurrences)
        .sort(([a, aCount], [b, bCount]) => bCount - aCount || parseInt(a) - parseInt(b))
        .map(([time, count]) => `Occurrence: ${count}    Time: ${time}`)
        .join('\n');

    await writeToFile('time_occurrences.txt', `Number of occurrences for each time (cutoff date is ${cutoffDate}):\n${byTime}`);
    await writeToFile('most_common_times.txt', `Times sorted by which is most common (cutoff date is ${cutoffDate}):\n${byOccurrence}`);
};

const main = async (cutoffDate = '2023-02-21') => {
    try {
        const filePath = path.join(__dirname, '..', 'text_files', 'cleaned_data.txt');
        const text = await fs.readFile(filePath, 'utf8');
        const sortedScores = processScores(text, cutoffDate);
        const anonSortedScores = anonymizeNames(sortedScores);
        const averageTimes = calculateAverageTimes(anonSortedScores);
        await analyzeAverageTimes(averageTimes, cutoffDate);
        const placements = analyzePlacementInfo(anonSortedScores);
        await findAveragePlace(placements);
        await findNumberOfFirsts(placements);
        await countTimesOccurrences(anonSortedScores, cutoffDate);
        console.log('Analysis complete. Check the text_files directory for results.');
    } catch (error) {
        console.error('An error occurred:', error);
    }
};

if (require.main === module) {
    const cutoffDate = process.argv[2] || undefined;
    main(cutoffDate);
}

module.exports = { main };  // Export for potential use in other modules